### PR TITLE
Implement Query API for prod_name, quantity and condition

### PR DESCRIPTION
**Pull request by majia67**
This PR implements the Query API. Comprehensive unit tests are included in the PR. The "list all" API is combined into the Query API due to their shared API paths.

Here's a screenshot:

image

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -210,6 +210,42 @@ def find(prod_id):
...
+            condition (string): the condition of the inventory you want to match
+        """
+        if condition == "new":
+            return ProductInformation.query.filter(ProductInformation.new_qty > 0).all()
+        elif condition == "used":
+            return ProductInformation.query.filter(ProductInformation.used_qty > 0).all()
+        elif condition == "open-boxed":
+            return ProductInformation.query.filter(ProductInformation.open_boxed_qty > 0).all()
+        else:
+            return None
```
[the remark this answers is not in the document]
Comment (jockyxu1122) [reply models.py]: Agreed

File: server.py [hunk 2]
```diff
@@ -113,12 +114,34 @@ def index():
...
 @app.route(PATH_INVENTORY, methods=[GET])
-def list_all_prod_info():
-    """ Return all entries in the Inventory system """
-    all_prod_info = ProductInformation.list_all()
+def query_prod_info():
+    """ Query specific entries in the Inventory system """
+    all_prod_info = []
+    if request.args.get('prod_name'):
+        prod_name = request.args.get('prod_name')
+        if prod_name == None:
```
Comment (jockyxu1122) [inline server.py]: This if block is never processed. If prod_name is None, it wouldn't enter if statement at line 120.

File: server.py [hunk 3]
```diff
@@ -113,12 +114,34 @@ def index():
...
+            abort(status.HTTP_400_BAD_REQUEST, INVALID_PARAMETER_MSG)
+        all_prod_info = ProductInformation.find_by_name(prod_name)
+    elif request.args.get('quantity'):
+        quantity = request.args.get('quantity')
+        try:
+            quantity = int(quantity)
```
Comment (jockyxu1122) [inline server.py]: Suggest using `if isinstance(quantity, int):`

File: tests/test_prod_info.py [hunk 4]
```diff
@@ -205,6 +205,52 @@ def test_deserialize_update_prod_info(self):
         prod_info = ProductInformation(prod_name = "ert", new_qty = -1, used_qty = -1, open_boxed_qty = -1, restock_level = -2, restock_amt = -1)
         self.assertRaises(DataValidationError, prod_info.deserialize_update, data)
 
+    def test_query_by_prod_name(self):
```
Comment (jockyxu1122) [inline tests/test_prod_info.py]: Naming issue. test_find_by_... to match tested function names.
line 222, 238 as well.

File: server.py (hunk 3 above)
Comment (jockyxu1122) [reply server.py]: Also, check for negative quantity request. Add tests correspondingly.

Comment (majia67) [reply server.py]: I think the args passed into the request are all strings by default?

Comment (majia67) [reply server.py]: I'm not sure if negative quantity is not allowed. My implementation could handle the negative quantity properly (i.e. returning an empty list if no record matches).

**Pull request merged**